rework console autocomplete sorting

Teero888 opened the pull request:
<!-- What is the motivation for the changes of this pull request? -->

<!-- Note that builds and other checks will be run for your change. Don't feel intimidated by failures in some of the checks. If you can't resolve them yourself, experienced devs can also resolve them before merging your pull request. -->

This PR changes the console autocompletion suggestions sorting to be more friendly and also refactors the code a bit to avoid code duplication.

------------------------------------------

Behavior before this PR:
 - filter by search query
 - alphabetical sorting (indirect)

This is very unintuitive and makes tabbing often completely useless because it is faster to type out the whole command instead of tabbing 7 times to get to the desired command:
<img width="2557" height="106" alt="image" src="https://github.com/user-attachments/assets/2a18a100-d1d6-4a12-bed4-bd404f4466ce" />
<img width="2500" height="101" alt="image" src="https://github.com/user-attachments/assets/15adf7c9-c271-44bb-bbeb-ff33d58ef5ab" />
^ I'm trying to get to `connect`

 ----------------------------------------------

Behavior after this PR:
- filter by search query
- sort based on the position of the search query within the suggestion string, in case of of a tie the shorter suggestion is preferred
<img width="2560" height="105" alt="image" src="https://github.com/user-attachments/assets/d30f8578-0dc0-4fb4-98ea-ab576391c8a9" />
<img width="2298" height="97" alt="image" src="https://github.com/user-attachments/assets/6ceddac3-f06f-4e8f-a358-4d2e36fc1140" />

------------------------------------------------

More examples:

Old:
<img width="2560" height="95" alt="image" src="https://github.com/user-attachments/assets/84c64600-80af-4180-92a3-13e2f6611699" />
New:
<img width="2560" height="97" alt="image" src="https://github.com/user-attachments/assets/3fcab8a6-f93e-4a88-a302-7419653bc131" />

Old:
<img width="2560" height="101" alt="image" src="https://github.com/user-attachments/assets/cbea6082-7137-45b2-b41f-a38180965c75" />
New:
<img width="2560" height="105" alt="image" src="https://github.com/user-attachments/assets/95514671-c65a-4c7d-b079-36e0752d7439" />



## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [x] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

AssassinTee commented on src/game/client/components/console.cpp:
```diff
@@ -1247,26 +1299,30 @@ void CGameConsole::OnRender()
-					else if(CompletionType == EArgumentCompletionType::KEY)
-						NumArguments = PossibleKeys(Info.m_pCurrentCmd, Input(), PossibleCommandsRenderCallback, &Info);
+
+					for(size_t i = 0; i < pConsole->m_vArgumentSuggestions.size(); ++i)
```
same here

Teero888 commented on src/game/client/components/console.cpp:
```diff
@@ -1247,26 +1299,30 @@ void CGameConsole::OnRender()
 			Info.m_Cursor.m_FontSize = FONT_SIZE;
-			const int NumCommands = m_pConsole->PossibleCommands(Info.m_pCurrentCmd, pConsole->m_CompletionFlagmask, m_ConsoleType != CGameConsole::CONSOLETYPE_LOCAL && Client()->RconAuthed() && Client()->UseTempRconCommands(), PossibleCommandsRenderCallback, &Info);
+
+			for(size_t i = 0; i < pConsole->m_vCommandSuggestions.size(); ++i)
```
oki thx <3

AssassinTee commented on src/game/client/components/console.cpp:
```diff
@@ -176,6 +176,31 @@ static int PossibleKeys(const char *pStr, IInput *pInput, IConsole::FPossibleCal
+	((std::vector<const char *> *)pUser)->push_back(pStr);
+}
+
+static void SortCompletions(std::vector<const char *> &vCompletions, const char *pSearch)
```
Thanks for the clarification

Robyt3 commented on src/game/client/components/console.cpp:
This does not preserve the scroll position for completion options while new options are being added. This previously mostly worked because options were only on added at the end of the list. Not sure if this is solvable with the new sorting though.

Robyt3 commented on src/engine/client/client.cpp:
```diff
@@ -1885,6 +1886,7 @@ void CClient::ProcessServerPacket(CNetChunk *pPacket, int Conn, bool Dummy)
 			if(!Unpacker.Error())
 			{
 				m_pConsole->RegisterTemp(pName, pParams, CFGFLAG_SERVER, pHelp);
+				((CGameClient *)GameClient())->m_GameConsole.ForceUpdateRemoteCompletionSuggestions();
```
Add a virtual function for this. The engine should not depend on the game implementation.

Teero888 commented:
> It crashed for me after typing `sv_map`, then selecting some completions and pressing enter to send the command.
> 
> A different time the map completions didn't work at all with either `sv_map` or `change_map`.
> 
> Crash log:
> 
> ```
> AddrPC           Params
> 00007FFF40A2DB24 0000000000000003 00007FF7A6342AA0 00007FF7A63429B2  ntdll.dll!NtTerminateProcess+0x14
> 00007FFF409EEF88 0000000000000003 00007FF7A6342AA0 0000006854F9D060  ntdll.dll!RtlExitUserProcess+0xb8
> 00007FFF3FA1E3EB 0000006854F9D040 0000000000000000 0000006854F9D150  KERNEL32.DLL!FatalExit+0xb
> 00007FFF3EA8A155 0000000000000001 000001E7F06646D0 0000000000000001  msvcrt.dll!_exit+0x85
> 00007FFF3EA8A7C5 00007FF7A63471F8 00007FF7A6342AA0 00007FF700000000  msvcrt.dll!_initterm_e+0x245
> 00007FFF3EA7F26D 0000006854F9D618 000001E7F2C37804 00007FF7A6342AA0  msvcrt.dll!abort+0x8d
> 00007FF7A62E7CEB 000001E7F2C37F00 000001E7F3016450 FFFFFFFF00000000  DDNet.exe!std::__glibcxx_assert_fail+0x7b
> 00007FF7A5C64C17 000001E7F2C37F00 0000000000000013 00007FFF00000004  DDNet.exe!std::vector<char const*, std::allocator<char const*> >::operator[]+0x57  [D:/Standalone-Software/msys64/mingw64/include/c++/15.2.0/bits/stl_vector.h @ 1263]
>   1261:       operator[](size_type __n) _GLIBCXX_NOEXCEPT
>   1262:       {
> > 1263: __glibcxx_requires_subscript(__n);
>   1264: return *(this->_M_impl._M_start + __n);
>   1265:       }
> 00007FF7A5DB7802 000001E7F2C37718 0000006800000003 0000000044160000  DDNet.exe!CGameConsole::CInstance::UpdateCompletionSuggestions+0x508  [ddnet/src/game/client/components/console.cpp @ 393]
>    391: }
>    392: }
> >  393: if(aOldArgument[0] != '\0' && str_comp(m_vpArgumentSuggestions[m_CompletionChosenArgument], aOldArgument) != 0)
>    394: {
>    395: for(size_t SuggestedId = 0; SuggestedId < m_vpArgumentSuggestions.size(); SuggestedId++)
> 00007FF7A5DBB975 000001E7F2C581C0 0000006854F9EA50 000001E73F008081  DDNet.exe!CGameConsole::OnRender+0x1111  [ddnet/src/game/client/components/console.cpp @ 1331]
>   1329: if(!pConsole->m_Searching && (m_ConsoleType == CONSOLETYPE_LOCAL || Client()->RconAuthed()) && !pConsole->m_Input.IsEmpty())
>   1330: {
> > 1331: pConsole->UpdateCompletionSuggestions();
>   1332: 
>   1333: CCompletionOptionRenderInfo Info;
> 00007FF7A5F36C8F 000001E7F207B040 0000000000000001 000000003B9ACA00  DDNet.exe!CGameClient::OnRender+0x32f  [ddnet/src/game/client/gameclient.cpp @ 834]
>    832: // render all systems
>    833: for(auto &pComponent : m_vpAll)
> >  834: pComponent->OnRender();
>    835: 
>    836: // clear all events/input for this frame
> 00007FF7A5CD130F 000001E7F207B040 000000684424BEA1 000000683F800000  DDNet.exe!CClient::Render+0x5d  [ddnet/src/engine/client/client.cpp @ 1121]
>   1119: else
>   1120: {
> > 1121: GameClient()->OnRender();
>   1122: }
>   1123: 
> 00007FF7A5CDA01C 0000006854F9F504 0000006854F9F510 0000000000000004  DDNet.exe!CClient::Run+0x1354  [ddnet/src/engine/client/client.cpp @ 3310]
>   3308: m_LastRenderTime = Now;
>   3309: 
> > 3310: Render();
>   3311: m_pGraphics->Swap();
>   3312: }
> 00007FF7A5CE2B11 00007FF700000001 00007FF7A6507CD8 0000000000000001  DDNet.exe!main+0x12a9  [ddnet/src/engine/client/client.cpp @ 4950]
>   4948: // run the client
>   4949: log_trace("client", "initialization finished after %.2fms, starting...", (time_get() - MainStart) * 1000.0f / (float)time_freq());
> > 4950: pClient->Run();
>   4951: 
>   4952: const bool Restarting = pClient->State() == CClient::STATE_RESTARTING;
> 00007FF7A5AF10C9 0000000000000000 0000000000000000 0000000000000000  DDNet.exe!__tmainCRTStartup+0xb9  [D:/W/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c @ 236]
> 00007FF7A5AF13F6 0000000000000000 0000000000000000 0000000000000000  DDNet.exe!WinMainCRTStartup+0x16  [D:/W/B/src/mingw-w64/mingw-w64-crt/crt/crtexe.c @ 91]
> 00007FFF3FA17374 0000000000000000 0000000000000000 0000000000000000  KERNEL32.DLL!BaseThreadInitThunk+0x14
> 00007FFF409DCC91 0000000000000000 0000000000000000 0000000000000000  ntdll.dll!RtlUserThreadStart+0x21
> ```

Thanks i completely forgot to update the sorting when new map suggestions come in. I was only calling `ForceUpdateConsoleRemoteCompletionSuggestions` when new normal commands were being sent so the map suggestions had dangling pointers. Fixed.
Also quick side note: We sadly can't just check if the size of the array is different from the last time we checked and update then because multiple elements could be removed and added inbetween OnRender calls.
